File: iq/14/intersection/find.cc
// Copyright 2025 Samvel Khalatyan. All rights reserved.
//
// clang-format off-next-line
//go:build ignore

#include "iq/14/intersection/find.h"

#include <memory>
#include <tuple>

namespace iq::intersection {
namespace {

struct Chain {
  std::shared_ptr<Node> head;
  std::shared_ptr<Node> tail;

  class ForwardIterator {
   public:
    ForwardIterator(Chain* chain, bool is_begin)
        : chain_(chain),
          is_head_(is_begin),
          node_(is_begin ? chain->head : nullptr) {}

    ForwardIterator& operator++() {
      if (node_) {
        node_ = node_->next;
        if (!node_ && is_head_) {
          is_head_ = false;
          node_ = chain_->tail;
        }
      }
      return *this;
    }

    const std::shared_ptr<Node>& operator*() const { return node_; }

    bool operator==(const ForwardIterator& that) const {
      if (this == &that) {
        return true;
      }
      return std::tie(chain_, is_head_, node_) ==
             std::tie(that.chain_, that.is_head_, that.node_);
    }

   private:
    const Chain* chain_;
    bool is_head_;
    std::shared_ptr<Node> node_;
  };

  ForwardIterator begin() { return ForwardIterator(this, true); }

  ForwardIterator end() { return ForwardIterator(this, false); }
};

std::shared_ptr<Node> find(Chain a, Chain b) {
  auto ait = a.begin();
  auto aend = a.end();
  auto bit = b.begin();
  auto bend = b.end();
  while (ait != aend && bit != bend) {
    if (*ait == *bit) {
      return *ait;
    }
    ++ait;
    ++bit;
  }
  return nullptr;
}

}  // namespace

std::shared_ptr<Node> Find(std::shared_ptr<Node> a, std::shared_ptr<Node> b) {
  return find(Chain(a, b), Chain(b, a));
}

}  // namespace iq::intersection

File: iq/14/intersection/find.go
// Copyright 2025 Samvel Khalatyan. All rights reserved.

package intersection

import (
	"fmt"
	"iter"
)

type Node struct {
	val  int
	next *Node
}

func NewList(nn ...int) *Node {
	var (
		head *Node
		tail *Node
	)
	for _, n := range nn {
		node := &Node{val: n}
		switch {
		case head == nil:
			head = node
		case tail != nil:
			tail.next = node
		}
		tail = node
	}
	return head
}

func (node *Node) String() string {
	nn := node.ToList()
	return fmt.Sprint(nn)
}

func (node *Node) ToList() []int {
	var nn []int
	for node != nil {
		nn = append(nn, node.val)
		node = node.next
	}
	return nn
}

func (node *Node) Append(tail *Node) *Node {
	head := node
	for node.next != nil {
		node = node.next
	}
	node.next = tail
	return head
}

func Find(a, b *Node) *Node {
	if a == nil || b == nil {
		return nil
	}
	for na, nb := range zip(join(a, b), join(b, a)) {
		if na == nb {
			return na
		}
	}
	return nil
}

func join(a, b *Node) iter.Seq[*Node] {
	return func(yield func(*Node) bool) {
		walk := func(n *Node) bool {
			for n != nil {
				if !yield(n) {
					return false
				}
				n = n.next
			}
			return true
		}
		if !walk(a) {
			return
		}
		if !walk(b) {
			return
		}
	}
}

func zip(a, b iter.Seq[*Node]) iter.Seq2[*Node, *Node] {
	return func(yield func(x, y *Node) bool) {
		nexta, stopa := iter.Pull(a)
		defer stopa()
		nextb, stopb := iter.Pull(b)
		defer stopb()
		for {
			x, xok := nexta()
			y, yok := nextb()
			if xok == false && yok == false {
				break
			}
			if !yield(x, y) {
				break
			}
		}
	}
}
